File: web/src/talk.js
// What players send each other alongside the cards, ported from the Android app's
// Talk.kt. None of it touches the rules, so none of it lives in the engine.

import { STICKER_ITEMS } from './cosmetics.js';

/**
 * A line has to fit in a data-channel frame, not in a novel. Long enough for a taunt,
 * short enough that nobody can stall the table by pasting a page of text.
 */
export const MAX_CHARS = 120;

/**
 * Every sticker there is, in catalogue order. Sent as an index, never as text: a build
 * that does not know a sticker ignores it instead of drawing a tofu box.
 *
 * The index is into the *whole* catalogue, not into what the sender has unlocked —
 * otherwise two players at different levels would disagree about what index 7 means.
 * Which of these the rail actually offers is a separate question, answered by the
 * sender's level.
 */
export const STICKERS = STICKER_ITEMS.map((item) => item.text);

export const sticker = (index) => STICKERS[index] ?? null;

/** Trims and caps a typed line; returns null when there is nothing worth sending. */
export function cleanLine(raw) {
  const text = String(raw ?? '').trim().slice(0, MAX_CHARS);
  return text.length ? text : null;
}

File: app/src/main/java/com/zknw/unoduo/net/Talk.kt
package com.zknw.unoduo.net

import com.zknw.unoduo.progress.Cosmetics

/**
 * What players throw at each other alongside the cards. None of it touches the rules, so
 * none of it lives in the engine or in a snapshot — it is relayed and forgotten.
 *
 * Stickers only, deliberately. Typing at a card table means looking away from it.
 */
object Talk {

    /**
     * Every sticker there is, in catalogue order. Sent as an index, never as text: one
     * byte on the wire, and a build that does not know a sticker simply ignores it
     * instead of drawing a tofu box.
     *
     * The index is into the *whole* catalogue, not into what the sender has unlocked —
     * otherwise two players at different levels would disagree about what index 7 means.
     * Which of these the rail actually offers is a separate question, answered by the
     * sender's level.
     */
    val STICKERS: List<String> = Cosmetics.stickers.map { it.text }

    fun sticker(index: Int): String? = STICKERS.getOrNull(index)
}
